---
title: "Labbook"
output:
  html_document: 
    keep_md: yes
    selfcontained: yes
    toc: yes
    toc_depth: 4
    toc_float: true
runtime: shiny
---

```{css, echo = FALSE}
/*from Eric koncina*/
.right-wborder{
  text-align: right;
  padding-right: 10px;
  padding-top: 5px;
  border: 1px solid black;
  border-radius: 10px;
  width: 300px;
  margin-left: auto;
}

```


<div class="right-wborder">

**Aurélien Ginolhac**  
Adjoint de Recherche  
University of Luxembourg  
Faculty of Science, Technology and Communication  
Life Science Research Unit  
Campus Belval, Biotech II – Office 4.22  
6, Avenue du Swing  
L-4367 Belvaux  
Phone: (+352) 46 66 44 6560  
aurelien.ginolhac@uni.lu  

</div>


## Project codes

### Research

6 numbers in the form _AACCNN_ where _AA_ is the year, _CC_ group code as below and _NN_ an increment number.

-----------------------------------------------------------------------
Team                                PI                          Code
----------------------------------- --------------------------  -------
SIGNAL TRANSDUCTION                 Iris BEHRMANN               01

CALCIUM SIGNALING AND INFLAMMATION  Eric TSCHIRHART             02

CYTOSKELETON                        Daniel ABANKWA              03

NEURO INFLAMMATION                  Paul HEUSCHLING             04

MOLECULAR DISEASE MECANISMS         Serge HAAN                  05

SYSTEMS BIOLOGY                     Thomas SAUTER               06
-----------------------------------------------------------------------

### Teaching

6 numbers in the form _AACCCC_ where _AA_ is the year, _CCCC_ code below

-----------------------------------------------------------
Title                         PI                      Code
----------------------------  ----------------------  -----
Biostatistic 2                Jean-Luc BUEB           22

Chip-seq doctoral school      Lasse Sinkkonen         61

R Workshop the LSRU/LCSB      Aurelien Ginolhac       33

Student internship            Alex Salsmann           41

HPC school, R session         Sébastien Varrette      31

-----------------------------------------------------------


## Project Overview

-----------------------------------------------------------------------------------------
Title                                       Contact                               Code
------------------------------------------- ------------------------------------- -------
IL6 and hypoxia in liver cancer cells,      Andreas Zimmer                        140101
expression data                             

Melanomics, NGS data, DNA-RNA               Susanne Reinsbach                     140102

Mass spec data, fold change                 Susann Murdich                        140201

Myosin and cancer, methylation and          Elisabeth Letellier                   140501
expression correlation

Heatmaps for selected expression data       Pit Ullmann                           140502

Myosin 5B annotation sequence               Martine Schmitz                       140503

Cluster of cancer genes and expression      Elisabeth Letellier                   140504

Liftover affymetrix Hugene 1.0 and 2.0      Maria Pacheco                         140601

Chip-seq analysis pipeline                  Lasse Sinkkonnen                      140602

Filter microarray data based                Pit Ullmann                           150501
on a certain threshold

Chip-seq analysis, time series              Déborah Gérard                        150601

Reliable marker for mature astrocyte        Luc Grandbarbe                        150401

Comparing replicate's means                 Maiti Lommel                          150301

Correlation between MYO5B expression        Pit Ullmann                           150502
 and miRNAs

Bubblechart with microRNA experiment        Maria Pacheco                         150603

Functional analysis of microarray data      Komal Baig                            150503

PCA and MDS of metabolites concentrations   Ingrid Gergei                         150604

Chip Seq analysis for Maria Pavlou / Lasse  Lasse Sinkkonen                       150605

Comparison of somatic mutations             Komal / Elisabeth                     150504

Survival curve on tissue microarray         Christelle Bahlawane                  150505

SNPs in Parkinson disease                   Lasse Sinkkonen                       150606

Whole exome sequencing melanoma             Giula Cesi / Ines                     150101

Migration / invasion curve                  Maiti Lommel                          150302

Heatmap for visualization of                Giula Cesi                            150102
exome sequencing

Statistical analysis of a time course       Véronique Schenten                    150201

Normalization of qPCR Ct values             Phuong Nguyen                         160601

Cross Combining strategy for PD disease     Lasse Sinkkonen                       160602

ISCU expression and survival in CAD         Pit Ullmann                           160501

Chip-seq analysis, PD mutations             Sarah Nickels / Lasse Sinkkonen       160603

liver gene / mir expression                 Florence / Melanie / Iris             160101

Linear modelling of RNA- and chip-seq       Julia Becker / Lasse Sinkkonen        160604

L-plasticin proteomics                      Elisabeth Reckinger-Schaffner         160301

mir 215 transfection                        Jil Weiler, Pit Ullmann               160502

Metabolites analysis, midica16              Dominik Ternes                        160503

Filter microarray                           Pit Ullmann                           160504

Autophagy                                   Komal Baig                            160505

Melanomics on patient 6                     Anke and Stephanie Kreis              170101

Chip-seq analysis SRA TH+ D50               Jochen Ochnmacht / Lasse Sinkkonen    170601

FASTCORMICS drugs capture                   Maria Pacheco                         170602

dataviz MTX and/or PDT                      Lasse / Nicole Rufo                   170603

HIF bindings                                Martine Schmitz                       170501

dataviz MTX and/or PDT                      Lasse                                 170604

Effects of stress on colon cancer           Martin Nurmik                         170502

RNA-seq and ChIP-seq mouse                  Yujuan Gui                            170605

Compare invasion curves                     Dunja Stojevski                       170301

RNA-seq human                               Tamara Bintener & Great Del Mistro    170606

DREAM challenge                             Sebastien De Landtsheer               170607

ALK / BRAF mutations                        Giulia Cesi                           170102

-----------------------------------------------------------------------------------------

## 2014

see [archive](http://shiny-server.uni.lu/2014.html)

## 2015

see [archive](http://shiny-server.uni.lu/2015.html)

## 2016

see [archive](http://shiny-server.uni.lu/2016.html)

## 2017

----------------------------------------------------------------------------------------------------
Date   Project Description
------ ------  -------------------------------------------------------------------------------------
170102 INFO    generalized labbook when year shifts


## TODO

- create R package for clustered heatmap `ggclust`. Clustering could be done by the [seriation package](https://cran.r-project.org/web/packages/seriation/index.html)? 
- create a skeleton Rmd for the labbook in `bifag`
- create `ggstars` skeleton in `Work/140504/diff_expression.Rmd`. Works with facet row OR col but not both. clean up use of global var in promises
- set up https on the shiny server, see help [here](http://www.r-bloggers.com/deploying-your-very-own-shiny-server/) and [recently](https://www.datascienceriot.com/add-authentication-to-shiny-server-with-nginx/kris/).
- count unnamed projects per team. `1606--` should mean an action for the Sauter group in 2016 that was not a dedicated project. 
Started writing this on 20161018


## Statistics


```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE)
library("tidyverse")
library("shiny")
ggplot2::theme_set(ggplot2::theme_bw(16))
library("lubridate")
library("timevis")
library("assertthat")
```


```{r koncina}
current_year <- format(Sys.time(), "%Y")
rmarkdown::pandoc_convert("labbook.Rmd", to = "html", output = paste0(current_year, ".html"))
```

### Count daily actions per year

```{r read_log, message = FALSE}
# load data
years <- seq(2014, current_year %>% as.numeric()) %>% as.character()
lab_tables <- map(years, ~ XML::readHTMLTable(paste0(.x, ".html"),
                                              colClasses = rep("character", 3),
                                              stringsAsFactors = FALSE,
                                              # only the daily actions
                                              which = 4)) %>%
  purrr::set_names(nm = years) %>%
  bind_rows(.id = "Year")

# sanity chech ####
assert_that(are_equal(colnames(lab_tables), c("Year", "Date", "Project", "Description"))) %>% message()

# test if all date are correctly ascending
my_date <- parse_date(lab_tables$Date, format("%y%m%d"))
if (diff(my_date) %>% min < 0) {
  data.frame(date = my_date, diff = c(NA, diff(my_date))) %>%
    filter(diff < 0) %>% print.data.frame
  stop("offending descending date", call. = FALSE)
}

```

```{r overview, fig.height = 9}

book <- lab_tables %>%
  mutate(Project = sub("^$", "unclassified", Project),
         Project = sub("^.{1}$", "unclassified", Project)) %>%
  filter(nchar(Date) == 6) %>%
  as_tibble()
# start shiny
selectInput("dataset", "Choose Dataset:", c("all", unique(book$Year)))
renderPlot({
  if (input$dataset == "all") {
    p <-   book %>%
  ggplot(aes(x = Project, fill = Year)) +
  geom_bar() +
  coord_flip() +
  ggtitle("all")
  } else {
    p <- book %>%
      filter(Year == input$dataset) %>%
      ggplot(aes(x = Project)) +
      geom_bar() +
      coord_flip() +
      ggtitle(input$dataset)
  }  
  plot(p)
}, height = 800, width = 700)

```


### Count daily action, all

**year** refer to the year when the project started

```{r team_plot, fig.height = 6, fig.width = 7}
# fetch the Team table
team_code <- XML::readHTMLTable(paste0(current_year, ".html"),
                                colClasses = rep("character", 3),
                                stringsAsFactors = FALSE,
                                # only the daily actions
                                which = 1)
# merge Team info to the labbook
book %>%
  separate(Project, c("year", "labcode"), sep = 2) %>%
  separate(labcode, c("Code", "nb"), sep = 2) %>%
  inner_join(team_code, by = "Code") %>%
  mutate(year = paste0("20", year)) -> book_team
book_team %>%
  ggplot(aes(x = Team, fill = year)) +
  geom_bar() +
  coord_flip() +
  xlab(NULL) +
  theme(axis.text.y = element_text(size = 8))
```

### Count project

Number of projects started at the corresponding **year** and filled by **Team**

```{r book_team, fig.height = 6, fig.width = 7}
book_team %>%
  select(year:nb, Team) %>%
  distinct() %>%
  ggplot(aes(x = year, fill = Team)) +
  geom_bar() +
  coord_flip() +
  xlab(NULL) +
  theme(axis.text.y = element_text(size = 8))
```


### Timeline

Average actions per week

```{r read_git, include = FALSE}
# QC from https://blog.rstudio.org/2016/08/05/readr-1-0-0/
#repo <- read_tsv("repo_date_stats.txt", col_types = cols(repo = col_character(),
#                                                         date = col_datetime(format = "%Y-%m-%d %H:%M:%S %z")))
# repo_spec <- write_rds(spec(repo), "repo-spec.rds")
repo <- read_tsv("repo_date_stats.txt", col_types = read_rds("repo-spec.rds"))
stop_for_problems(repo)
if (nrow(repo) == 0) stop("git log is empty", call. = FALSE)
repo %>%
  replace_na(list(file_changed = 0, insertions = 0, deletions = 0)) %>%
  mutate(deletions = -deletions) %>%
  filter(deletions > -2000, insertions < 2000) -> repo
```


```{r git_labbook, fig.width = 10, fig.height = 7, warning = FALSE}
book %>%
  mutate(date = ymd(Date)) %>%
  mutate(week = week(date),
         month = month(date),
         year = year(date)) %>%
  group_by(year, week) %>%
  mutate(week_count = n()) %>% 
  group_by(year, month) %>%
  mutate(mean = mean(week_count),
         sd = sd(week_count)) %>%
  dplyr::select(year, month, week, week_count, mean) %>%
  ungroup() %>%
  distinct() -> timeline_book

repo %>%
  filter(author == "ginolhac@gmail.com") %>%
  mutate(week = week(date),
         month = month(date),
         year = year(date)) %>%
  count(year, month, week) %>%
  ungroup() %>%
  full_join(timeline_book,  by = c("year", "month", "week")) %>%
  ggplot(aes(x = month)) +
  stat_summary(aes(y = n, colour = "git (mean+se commit)"), fun.data = "mean_se", na.rm = TRUE) +
  geom_line(aes(y = mean, colour = "labbook (mean)")) +
  geom_point(aes(y = mean, colour = "labbook (mean)"), na.rm = TRUE) +
  theme(legend.position = c(0.01, 0.98),
        legend.justification = c(0, 1),
        panel.grid.minor = element_blank()) +
  facet_wrap(~ year) +
  scale_x_continuous(breaks = seq(1, 12, 1)) +
  labs(y = NULL,
       colour = NULL)
```



###  In `git` repositories

insertions and deletions, discarding commit with > 2,000 lines and **biostat2**.

#### Overview 
```{r git_overview, fig.height = 8, fig.width = 10}


repo %>%
  filter(author == "ginolhac@gmail.com") %>%
  group_by(repo) %>%
  summarise(ins = sum(insertions),
            del = sum(deletions),
            n   = n(),
            n_del = sum(deletions < 0),
            n_ins = sum(insertions > 0)) %>%
  separate(repo, c("year", "labcode"), sep = 2, remove = FALSE) %>%
  separate(labcode, c("Code", "nb"), sep = 2) %>%
  inner_join(team_code, by = "Code") -> repo_ref
repo_ref %>%
  ggplot(aes(colour = Team)) +
  geom_hline(yintercept = 0, linetype = "dashed", colour = "grey50") +
  geom_segment(aes(x = repo, xend = repo, y = del, yend = ins)) +
  geom_point(aes(x = repo, y = del, size = n_del)) +
  geom_point(aes(x = repo, y = ins, size = n_ins)) +
  scale_size_area("# commits") +
  scale_colour_hue(labels = stringr::str_wrap(repo_ref$Team %>% factor %>% levels, width = 20)) +
  annotate("text", x = 0.8, y = -2000, label = "deletions") +
  annotate("text", x = 0.8, y = 10000, label = "insertions") +
  coord_flip() + xlab(NULL) + ylab("# lines")
```

#### Detail per project 

```{r git_details, fig.width = 12}
selectInput("project", "Choose project:", unique(repo$repo))
# plotting per project
renderPlot({
  repo %>%
    filter(repo == input$project) %>%
    #filter(repo == "140101") %>% 
    mutate(date = lubridate::date(date)) %>%
    ggplot(aes(x = date)) +
    geom_ribbon(aes(ymin = 0, ymax = insertions, fill = "insertions")) +
    geom_point(aes(y = insertions, size = file_changed, colour = "insertions")) +
    geom_ribbon(aes(date, ymin = deletions, ymax = 0, fill = "deletions")) +
    geom_point(aes(y = deletions, size = file_changed, colour = "deletions")) +
    scale_fill_manual(name = NULL, values = c("red", "green")) +
    scale_colour_manual(name = NULL, values = c("red", "green")) +
    scale_size("# files", range = c(0.5, 3), breaks = c(1, 3, 6, 9)) +
    scale_x_date() +
    facet_wrap(~ repo) +
    xlab(NULL) + ylab("# lines")
})
```

#### Timeline

```{r timeline}
# tweaked from https://ottlngr.github.io/posts/gitlog2timevis.html
selectInput("project_tv", "Choose project:", unique(repo$repo))
# plotting per project
renderTimevis({
  repo %>%
    filter(repo == input$project_tv) %>%
    mutate(type = "box",
           content = gsub("-", " ", content),
           id = seq_along(content)) %>%
    select(id, content, start = date, type) %>%
    distinct(start, content) %>%
    timevis()
})
```





#### Punchcard
```{r punch_card, fig.height = 9, fig.width = 13, warning = FALSE}

library(forcats)

repo %>%
  filter(author == "ginolhac@gmail.com") %>%
  select(repo, date) %>%
  mutate(day_of_week = wday(date, label = TRUE, abbr = FALSE),
         day_of_week = fct_relevel(day_of_week, "Sunday", after = Inf),
         hour = hour(date)) %>%
  count(day_of_week, hour) %>%
  complete(day_of_week, hour) %>%
  ggplot(aes(x = hour + 0.5, y = 1, size = n)) +
  geom_point() +
  facet_wrap(~ day_of_week, ncol = 1, scales = "free_x") +
  scale_size_area(max_size = 9) +
  scale_y_discrete() +
  theme_bw(16) +
  theme(strip.placement = "outside",
        strip.text.x = element_text(hjust = 0.8, size = 17),
        strip.background = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        axis.ticks.length = unit(-0.25, "cm"),
        axis.ticks = element_line(, colour = "grey60"),
        axis.text.x = element_text(margin = margin(t = 0.5, unit = "cm"), colour = "grey60"),
        plot.subtitle = element_text(face = "italic")) +
  scale_x_time(breaks = 1:24, labels = paste0(1:24, "h")) +
  labs(x = NULL,
       y = NULL,
       size = "count",
       title = "Punch card",
       subtitle = "commits' count")
  
  
```
